src/main/java/cloud/eppo/ConfigurationWrangler.java

```diff
@@ -0,0 +1,241 @@
+      }
+    }
+
+    setConfig(requestor.load(config));
+  }
+
+  // Load Async Method
+  // Start an async load via the configStore (which can be subclassed by libraries using this
+  // package to load from disk/cache/etc)
+  // Also start an async http fetch (this is required for Android to have network calls off the main
+  // thread, which is a good idea anyway).
+  //
+  // The precedence strategy is first-to-succeed, which is to say, the first loader to succeed gets
+  // to call
+  // `callback.onComplete()`.
+  //
+  // If the cache load succeeds, it writes the config iff the fetch has not yet set a value
+  // If the cache load fails, it calls `onFailure` iff the fetch also failed. Only the fetch error
+  // message is ever returned.
+  //
+  // If the fetch succeeds, it writes its config regardless of the cache result
+  // The fetch success will trigger the callback only if it has not yet been called.
+  // If the fetch fails and the cache failed pass the fetch error message to the callback
+  // (`onFailure`)
+  void loadAsync(LoadCallback callback) {
+    loadAsync(false, callback);
+  }
+
+  void loadAsync(final boolean skipCache, LoadCallback callback) {
+    if (skipCache) {
+      loadAsyncFromRequestor(callback);
+      return;
+    }
+
+    // We have two at-bats to load the configuration: loading from cache and fetching
+    // The below variables help them keep track of each other's progress
+    AtomicBoolean cacheLoadInProgress = new AtomicBoolean(true);
+    AtomicReference<String> fetchErrorMessage = new AtomicReference<>(null);
+    // We only want to fire the callback off once; so track whether or not we have yet
+    AtomicBoolean callbackCalled = new AtomicBoolean(false);
+
+    configurationStore.load(
+        new IConfigurationStore.CacheLoadCallback() {
+          @Override
+          public void onSuccess(Configuration result) {
+            lock.lock();
+            try {
+              cacheLoadInProgress.set(false);
+              // If cache loaded successfully, fire off success callback if not yet done so by
+              // fetching
+              if (callback != null && callbackCalled.compareAndSet(false, true)) {
+                log.debug("Initialized from cache");
+                callback.onSuccess(null);
+              }
+            } finally {
+              lock.unlock();
+            }
+          }
+
+          @Override
+          public void onFailure(String errorMessage) {
+            lock.lock();
+            try {
+              cacheLoadInProgress.set(false);
+              log.debug("Did not initialize from cache");
+              // If cache loading failed, and fetching failed, fire off the failure callback if not
+              // yet done so
+              // Otherwise, if fetching has not failed yet, defer to it for firing off callbacks
+              if (callback != null
+                  && fetchErrorMessage.get() != null
```

Comment:
I don't think we need to make sure `fetchErrorMessage` is non-null here. We should only be checking if a callback is at play in case something else goes wonky

Reply:
The `fetchErrorMessage` + callbackCalled.getAndSet are a sufficient proxy for something like a `fetchCallInProgress` boolean and are variables that we need to track anyway. An additional atomicBoolean would be more clear, if a little redundant.
Happy to change it, but I thought your original comments and code were sufficient here.